pkg/metadata/service_discovery.go

```diff
@@ -15,21 +15,54 @@
 package metadata
 
 import (
+	"errors"
 	"sync"
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/prometheus/common/model"
 
 	"github.com/parca-dev/parca-agent/pkg/discovery"
+	"github.com/parca-dev/parca-agent/pkg/process"
 )
 
+type serviceDiscoveryProvider struct {
+	StatefulProvider
```

Comment:
I find this hides a lot of context, I would prefer if we got rid of this and StatelessProvider entirely and rather have tiny amounts of duplication than this abstraction that hides very important context.

Reply:
Just for clarity, I'm fine with just not using StatefulProvider in the process discovery for this PR and we do follow ups for the other providers.

Reply:
👍 Sounds good. It's only used here. I'll do the refactoring.